temporalio/converter.py

```diff
@@ -0,0 +1,410 @@
+        except google.protobuf.message.DecodeError as err:
+            raise RuntimeError("Failed parsing") from err
+
+
+class JSONPlainPayloadConverter(PayloadConverter):
+    """Converter for 'json/plain' payloads supporting common Python values.
+
+    This supports all values that :py:func:`json.dump` supports and also adds
+    encoding support for :py:mod:`dataclasses` by converting them using
+    :py:func:`dataclasses.asdict`. Note that on decode, if there is a type hint,
+    it will be used to construct the data class.
+    """
+
+    _encoder: Optional[Type[json.JSONEncoder]]
+    _decoder: Optional[Type[json.JSONDecoder]]
+    _encoding: str
+
+    def __init__(
+        self,
+        *,
+        encoder: Optional[Type[json.JSONEncoder]] = None,
+        decoder: Optional[Type[json.JSONDecoder]] = None,
+        encoding: str = "json/plain",
+    ) -> None:
+        """Initialize a JSON data converter.
+
+        Args:
+            encoder: Custom encoder class object to use.
+            decoder: Custom decoder class object to use.
+            encoding: Encoding name to use.
+        """
+        super().__init__()
+        self._encoder = encoder
+        self._decoder = decoder
+        self._encoding = encoding
+
+    @property
+    def encoding(self) -> str:
+        """See base class."""
+        return self._encoding
+
+    async def encode(self, value: Any) -> Optional[temporalio.api.common.v1.Payload]:
+        """See base class."""
+        if dataclasses.is_dataclass(value):
+            value = dataclasses.asdict(value)
+        # We swallow JSON encode error and just return None
+        try:
+            return temporalio.api.common.v1.Payload(
+                metadata={"encoding": self._encoding.encode()},
+                data=json.dumps(
+                    value, cls=self._encoder, separators=(",", ":"), sort_keys=True
+                ).encode(),
+            )
+        except (RuntimeError, TypeError, ValueError):
+            return None
+
+    async def decode(
+        self,
+        payload: temporalio.api.common.v1.Payload,
+        type_hint: Optional[Type] = None,
+    ) -> Any:
+        """See base class."""
+        try:
+            obj = json.loads(payload.data, cls=self._decoder)
+            # If the object is a dict and the type hint is present for a data
+            # class, we instantiate the data class with the value
+            if (
+                isinstance(obj, dict)
+                and inspect.isclass(type_hint)
+                and dataclasses.is_dataclass(type_hint)
+            ):
+                obj = type_hint(**obj)
+            return obj
+        except json.JSONDecodeError as err:
+            raise RuntimeError("Failed parsing") from err
+
+
+# TODO(cretz): Should this be a var that can be changed instead? If so, can it
+# be replaced _after_ client creation? We'd just have to fallback to this
+# default at conversion time instead of instantiation time.
```

Comment:
If we do that, IMO it'd make sense to keep this and also introduce some kind of `set_default` so that you don't wipe out the original defaults and they're always easily available if you want to restore them for some reason.

Reply:
Yeah, I think I'd like to make it an immutable default. We have ways to set the default at the client level which then applies to all things within anyways. You don't need to set the "default default" I don't think (we don't allow it in Go).

Reply:
The other SDKs have an immutable default converter, but I see Spencer's point about setting the "default default" being more convenient than having to pass data converters everywhere.

Reply:
:+1: Keeping it immutable here. I have exposed converters from the default for easy creation of a new one.